posttroll/subscriber.py

```diff
@@ -413,3 +425,54 @@ def handle_msg(self, msg):
         else:
             LOGGER.debug("Removing address %s", str(addr_))
             self.subscriber.remove(addr_)
+
+
+def dict_config(settings):
+    """Get a subscriber class instance defined by a dictionary of configuration options.
+
+    The subscriber is created based on the given options in the following way:
+
+    - setting *settings['addresses']* to a non-empty list AND *settings['nameserver']* to *False*
+      will disable nameserver connections and connect only the listed addresses
+
+    - setting *settings['nameserver']* to a string will connect to nameserver
+      running on the indicated server
+
+    - if *settings['nameserver']* is missing or *None*, the nameserver on localhost is assumed.
+
+    Additional options are described in the docstrings of the respective classes, namely
+    :class:`~posttroll.subscriber.Subscriber` and :class:`~posttroll.subscriber.NSSubscriber`.
+
+    """
+    if settings.get('addresses') and settings.get('nameserver') is False:
+        return _get_subscriber_instance(settings)
+    return _get_nssubscriber_instance(settings)
+
+
+def _get_subscriber_instance(settings):
+    addresses = settings['addresses']
+    topics = settings.get('topics', '')
+    message_filter = settings.get('message_filter', None)
+    translate = settings.get('translate', False)
+
+    return Subscriber(addresses, topics=topics, message_filter=message_filter, translate=translate)
+
+
+def _get_nssubscriber_instance(settings):
+    services = settings.get('services', '')
+    topics = settings.get('topics', _MAGICK)
+    addr_listener = settings.get('addr_listener', False)
+    addresses = settings.get('addresses', None)
+    timeout = settings.get('timeout', 10)
+    translate = settings.get('translate', False)
+    nameserver = settings.get('nameserver', 'localhost') or 'localhost'
+
+    return NSSubscriber(
+        services=services,
+        topics=topics,
+        addr_listener=addr_listener,
+        addresses=addresses,
+        timeout=timeout,
+        translate=translate,
+        nameserver=nameserver
+    )
```

Comment:
how about making these class method call eg `from_dict_config`?

Reply:
Umm, what? I don't understand.

Reply:
instead of calling `_get_substriber_instance`, you would call `Subscriber.from_dict_config`. This can be implemented lik this:

```
# In the Subscriber class definition

@classmethod
def from_dict_config(cls, settings):
    addresses = settings['addresses']
    topics = settings.get('topics', '')
    message_filter = settings.get('message_filter', None)
    translate = settings.get('translate', False)

    return cls(addresses, topics=topics, message_filter=message_filter, translate=translate)

```

Reply:
Testing that only relevant arguments are passed to the class init seems to be getting too complicate, so I'll keep these as separate methods.

Reply:
Can we make the tests simpler?

Reply:
These tests are [as simple as they can be](https://github.com/pytroll/posttroll/blob/4ac8d98498089a985f26a6b339cc826dfb3f2c75/posttroll/tests/test_pubsub.py#L573-L623), but get overly complex if I'd use the version with class methods.

Reply:
Ok, not necessarily a relevant input to the problem at hand, sorry for digressing: I don't really like these tests :) because
the only this that is tested is that something like `Subscriber(some_arg, some_other_arg)` is written in the code. To me, this is not good because:
1. it doesn't break if the signature of `Subscriber` changes
2. IMO, here we test the implementation, while we should test the behaviour.

Would it be possible instead to check that passing a given argument has the desired effect? For example that  changing `topics` actually changes the subject of the published messages?

Sorry for being a stickler here, but I think this is important. Also, I realise that I have done this myself in the past, and you would be totally entitled to shout at me for that, but I'd like to gradually switch to better testing, and this is a good occasion to enforce a good behaviour. Hope it makes sense.

Reply:
1. The tests break if kwargs are removed or added in either `Subscriber` or `NSSubscriber`
2. I'm testing this implementation to behave as expected, which is that it is passing the correct args/kwargs to the different subscriber classes.The behaviour of these classes are tested in other test cases. The `create_*_from_dict_config()` are tested end-to-end through `TestNS::test_pub_sub_ctx` TestNS::test_pub_sub_add_rm`, `TestNSWithoutMulticasting::test_pub_sub_ctx`, `TestNSWithoutMulticasting::test_pub_sub_add_rm` and `TestNSWithoutMulticasting::test_pub_sub_ctx_no_nameserver`.

Reply:
1. if the names of the kwargs in the `Subscriber` or `NSSubscriber` constructors changes, the tests you are referring to don't catch it, right?

Reply:
Damn, that's right. How about I change the tests so that I don't mock them, but neither do I assert anything. That way we see that the kwargs truly match. I'll just mock `NSSubscriber.start()` so that it isn't started for nothing.

Reply:
That wouldn't catch if new kwargs were added, but better that than the current version of the tests...

Reply:
First iteration pushed.

Reply:
2. indeed, since the replacement of the calls to the `NSSubscriber` with `create_subscriber_from_dict_config`, the latter is tested end-to-end in behaviour with the tests you mention in your last comment. That means that the new/refactored code is effectively tested already.
The other tests that you mentionned earlier still only test implementation IMO. They rely on the fact that the new functions end up calling a given constructor, while we might change that to a unpickling (yaml loading) or class methods in the future. In other words, according to me, these tests are not useful and should be deleted.

That would then open for having class methods as mentioned earlier.